app/models/program.rb

```diff
@@ -16,5 +16,52 @@ class Program < ApplicationRecord
 
   validates :facility_code, presence: true
   validates :program_type, inclusion: { in: InstitutionProgram::PROGRAM_TYPES }
-  validates_with ProgramValidator, on: :after_import
+
+  def self.after_import_batch_validations(failed_instances)
+    duplicate_facility_description_results.each do |record|
+      message = line_number(record['csv_row']) + non_unique_error_msg(record['facility_code'], record['description'])
+      warning = { index: record['csv_row'], message: message }
+      failed_instances << warning
+    end
+
+    missing_facility_in_weam.each do |program|
+      message = line_number(program.csv_row) + BaseValidator.missing_facility_error_msg(program)
+      warning = { index: program.csv_row, message: message }
+      failed_instances << warning
+    end
+  end
+
+  def self.duplicate_facility_description_results
+    str = <<-SQL
+     SELECT programs.csv_row, programs.facility_code, programs.description
+      FROM programs
+        INNER JOIN (
+        SELECT
+          UPPER(facility_code) facility_code,
+          UPPER(description) description
+        FROM programs
+        GROUP BY
+          UPPER(facility_code),
+          UPPER(description)
+        HAVING COUNT(*) > 1
+        ) dupes on UPPER(programs.facility_code) = dupes.facility_code
+        AND UPPER(programs.description) = dupes.description
+    SQL
+
+    sql = Program.send(:sanitize_sql, [str])
+    Program.connection.execute(sql)
```

Comment:
We can make this more Rails-y too like so:
```ruby
subquery = Program.select('UPPER(facility_code) as facility_code, UPPER(description) as description').group('UPPER(facility_code), UPPER(description)').having('count(*) > 1')
Program.joins("INNER JOIN (#{subquery.to_sql}) dupes on UPPER(programs.facility_code) = dupes.facility_code AND UPPER(programs.description) = dupes.description")
```
And you'll have `Program` objects returned to use above instead of hashes.

Reply:
changed